java-connectors/kafka-connect-azure-eventhubs/src/test/java/io/lenses/streamreactor/connect/azure/eventhubs/source/EventHubsKafkaConsumerControllerTest.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2017-2024 Lenses.io Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.azure.eventhubs.source;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.lenses.streamreactor.connect.azure.eventhubs.config.SourceDataType;
+import io.lenses.streamreactor.connect.azure.eventhubs.config.SourceDataType.KeyValueTypes;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class EventHubsKafkaConsumerControllerTest {
+
+  private static final String INPUT_TOPIC = "INPUT";
+  private static final String OUTPUT_TOPIC = "OUTPUT";
+  private static final String INPUT_TOPIC_2 = "INPUT2";
+  private static final String OUTPUT_TOPIC_2 = "OUTPUT2";
+  public static final int DEFAULT_CAPACITY = 10;
+  private ArrayBlockingQueue<ConsumerRecords<byte[], byte[]>> recordsQueue;
+
+  private EventHubsKafkaConsumerController testObj;
+
+
+  @BeforeEach
+  void setUp() {
+    recordsQueue = new ArrayBlockingQueue<>(DEFAULT_CAPACITY);
+  }
+
+  @Test
+  void pollShouldPollQueueAndReturnSourceRecords() throws InterruptedException {
```

Comment:
Let's break this down into smaller methods so it's easier to read.

Reply:
Unfortunately I do not think this is a good idea here: unlikely other testcases that you've pointed in review those two that reside here are rather different so not that much can be extracted that is similar between those.

Also: there's a fine line when making test more readable, I wouldn't like to _overengineer_ it :)

Reply:
Firstly, if you decline to make changes please don't resolve the conversation.

I understand your concern, but breaking down a long test into smaller methods isn't overengineering; it's a best practice in software development. Smaller methods make code more readable, maintainable, and reusable.

While the two test cases you mentioned might seem different, there's often more commonality than meets the eye. Even if the immediate similarities aren't apparent, breaking down the test can reveal underlying patterns that can be extracted into reusable components. This not only improves readability but also reduces duplication and makes future maintenance easier.  It's a practice that benefits not only the current codebase but also future developers who will work on it.

Regarding the risk of overengineering, it's important to strike a balance. Overengineering occurs when unnecessary complexity is introduced, but breaking down a long test into smaller methods is a fundamental technique for improving code quality. It's about finding the right level of abstraction to make the code understandable without adding unnecessary complexity.

Reply:
I agree 💯 %! That's why I addressed every other testcase and I really like how some of them look (KCQL utility looks much better now). I was just under impression that here the tests are distinct enough to not extract them to helper methods, especially because of constants that are being put in place (helper methods will just hide them but the constants will suddenly appear in assertions). I changed it a bit to help with `ConsumerRecords` creation, but I think further changes may blur the image. If you think it's a blocker: I will change it, just let me know what exactly, but I wanted to highlight that those two testcases aren't the same as the ones before.

Reply:
Merged 1170 into this. Thanks!